src/calls/embedTemplatePage.ts

```diff
@@ -121,3 +111,98 @@ function createEmbedIframe(options: CreateIframeOptions): HTMLIFrameElement {
 
   return iframe;
 }
+
+function parseTokenPayload(token: string): SdkTokenPayload {
+  const tokenPayloadResult = sdkTokenPayloadSchema.safeParse(
+    JSON.parse(atob(token.split(".")[1] ?? "")),
+  );
+
+  if (!tokenPayloadResult.success) {
+    throw new Error("Invalid token payload");
+  }
+
+  return tokenPayloadResult.data;
+}
+
+function validatePermissions(
+  permissions: SdkPermission[],
+  features?: TemplateEmbedFeatures,
+): void {
+  const hasPermission = (validPermissionSets: SdkPermission[][]) => {
+    return validPermissionSets.every((permissionSet) =>
+      permissionSet.some((permission) => permissions.includes(permission)),
+    );
+  };
+
+  if (!hasPermission(REQUIRED_PERMISSIONS.base)) {
+    throw new Error(
+      "workspace:read, user:read, and template:read permissions are required",
+    );
+  }
+
+  if (!features) {
+    return;
+  }
+
+  if (features.passList && !hasPermission(REQUIRED_PERMISSIONS.passList)) {
+    throw new Error("passList feature requires pass:read permission");
+  }
+
+  if (
+    features.templateEditor &&
+    !hasPermission(REQUIRED_PERMISSIONS.templateEditor)
+  ) {
+    throw new Error(
+      "templateEditor feature requires template:write permission",
+    );
+  }
+
+  if (features.campaigns && !hasPermission(REQUIRED_PERMISSIONS.campaigns)) {
+    throw new Error("campaigns feature requires campaign:read permission");
+  }
+
+  if (
+    features.campaignEditor &&
+    !hasPermission(REQUIRED_PERMISSIONS.campaignEditor)
+  ) {
+    throw new Error(
+      "campaignEditor feature requires campaign:write permission",
+    );
+  }
+}
+
+type SdkPermission = z.infer<typeof sdkPermissionSchema>;
+const sdkPermissionSchema = z.enum([
+  "workspace:read",
+  "workspace:write",
+  "template:read",
+  "template:write",
+  "pass:read",
+  "pass:write",
+  "campaign:read",
+  "campaign:write",
+  "user:read",
+  "user:write",
+]);
+
+type SdkTokenPayload = z.infer<typeof sdkTokenPayloadSchema>;
+const sdkTokenPayloadSchema = z.object({
+  workspaceHandle: z.string(),
+  permissions: z.array(sdkPermissionSchema),
+});
+
+// Each feature requires at least one permission from each inner permission set
+const REQUIRED_PERMISSIONS: Record<
+  keyof TemplateEmbedFeatures | "base",
+  SdkPermission[][]
+> = {
+  base: [
+    ["workspace:read", "workspace:write"],
+    ["user:read", "user:write"],
+    ["template:read", "template:write"],
+  ],
```

Comment:
This could be a separate `const`. Maybe`REQUIRED_PERMISSIONS` and `REQUIRED_PERMISSIONS_BY_FEATURE`.